### PR TITLE
Added nifm service

### PR DESCRIPTION
**Pull request by shadowninja108**
* Added nifm service
* Implemented gethostid in /runtime/devices/socket
     * This is required to get the IP address of the Switch, as it would originally return 0.0.0.0

### CODE REVIEW

File: nx/source/runtime/devices/socket.c [hunk 1]
```diff
@@ -1580,6 +1581,29 @@ int getnameinfo(const struct sockaddr *sa, socklen_t salen,
     return gaie;
 }
 
+
+long gethostid(void) {
+	Result rc;
+	IGeneralService igs;
+	rc = CreateGeneralService(&igs);
```
Comment (roblabla) [inline nx/source/runtime/devices/socket.c]: You leak a handle here, by creating an instance of IGeneralService but never closing it.

**Comment by shadowninja108** [comment]
Fixed the service leak by creating the service upon init, then caching it.

File: nx/include/switch/services/nifm.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+#include "../services/sm.h"
+
+typedef struct {
+    Service  s;
+} IGeneralService;
+
+Result nifmInitialize();
+void nifmExit(void);
+
+Result _CreateGeneralService(IGeneralService* out, u64 in);
```
Comment (yellows8) [inline nx/include/switch/services/nifm.h]: Internal funcs shouldn't be in this .h.

File: nx/include/switch/services/nifm.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+Result _CreateGeneralServiceOld(IGeneralService* out);
+
+Result GetCurrentIpAddress(u32* out);
```
Comment (yellows8) [inline nx/include/switch/services/nifm.h]: nifm prefix missing.

File: nx/source/runtime/devices/socket.c [hunk 4]
```diff
@@ -1580,6 +1581,24 @@ int getnameinfo(const struct sockaddr *sa, socklen_t salen,
     return gaie;
 }
 
+
+long gethostid(void) {
+	Result rc;
```
Comment (yellows8) [inline nx/source/runtime/devices/socket.c]: tabs/spaces mixing + elsewhere

File: nx/include/switch/runtime/devices/socket.h [hunk 5]
```diff
@@ -21,6 +21,11 @@ typedef struct  {
     int dns_timeout;                            ///< For DNS requests: timeout or 0.
 } SocketInitConfig;
 
+long gethostid(void);
```
Comment (yellows8) [inline nx/include/switch/runtime/devices/socket.h]: Remove these.

File: nx/source/services/nifm.c [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+ * @brief Network interface service IPC wrapper.
+ * @author shadowninja108
+ * @copyright libnx Authors
+ */
+#include "services/nifm.h"
+
+static Service g_nifmSrv;
+static IGeneralService g_nifmIGS;
+
+Result nifmInitialize(){
```
Comment (yellows8) [inline nx/source/services/nifm.c]: refcounting missing

File: nx/source/services/nifm.c [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+}
+
+void nifmExit(void){
+	if(serviceIsActive(&g_nifmIGS.s))
+		serviceClose(&g_nifmIGS.s);
+	if(serviceIsActive(&g_nifmSrv))
+		serviceClose(&g_nifmSrv);
+}
+
+Result _CreateGeneralService(IGeneralService* out, u64 in){
```
Comment (yellows8) [inline nx/source/services/nifm.c]: static Result _nifm<...>

File: nx/source/services/nifm.c [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+        rc = resp->result;
+
+        if (R_SUCCEEDED(rc)) 
+            serviceCreate(&out->s, r.Handles[0]);      
+    }
+	
+	return rc;
+}
+
+Result _CreateGeneralServiceOld(IGeneralService* out){
```
Comment (yellows8) [inline nx/source/services/nifm.c]: see above

File: nx/source/services/nifm.c (hunk 6 above)
Comment (fincs) [reply nx/source/services/nifm.c]: Missing (void)

File: nx/include/switch/services/nifm.h [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+#include "../kernel/ipc.h"
+#include "../kernel/detect.h"
+
+#include "../services/sm.h"
+
+typedef struct {
+    Service  s;
+} IGeneralService;
+
+Result nifmInitialize();
```
Comment (fincs) [inline nx/include/switch/services/nifm.h]: Missing (void)

File: nx/source/runtime/devices/socket.c [hunk 10]
```diff
@@ -1596,15 +1615,3 @@ void setnetent(int a) { (void)a;}
 void setprotoent(int a) { (void)a; }
 
 /************************************************************************************************************************/
-
```
Comment (yellows8) [inline nx/source/runtime/devices/socket.c]: If this block is going to be moved the above comment line should be removed.

**Comment by shadowninja108** [comment]
I believe I've sorted every issue you brought up.

**Comment by shadowninja108** [comment]
Sorry, used an online code "beautifier", this should be better.

File: nx/source/services/nifm.c [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+
+static Result _nifmCreateGeneralService(IGeneralService* out, u64 in);
+static Result _nifmCreateGeneralServiceOld(IGeneralService* out);
+
+Result nifmInitialize(void) {
+    atomicIncrement64(&g_refCnt);
+
+    if (serviceIsActive(&g_nifmSrv))
+        return 0;
+    Result rc;
```
Comment (yellows8) [inline nx/source/services/nifm.c]: Add a newline here.

File: nx/source/runtime/devices/socket.c (hunk 4 above)
Comment (yellows8) [reply nx/source/runtime/devices/socket.c]: This still has tabs...

**Comment by shadowninja108** [comment]
Sorry, tabs are a force of habit

File: nx/source/runtime/devices/socket.c [hunk 12]
```diff
@@ -1580,6 +1581,24 @@ int getnameinfo(const struct sockaddr *sa, socklen_t salen,
     return gaie;
 }
 
+
```
Comment (yellows8) [inline nx/source/runtime/devices/socket.c]: extra newline

File: nx/source/services/nifm.c [hunk 13]
```diff
@@ -0,0 +1,154 @@
...
+ * @file nifm.h
+ * @brief Network interface service IPC wrapper.
+ * @author shadowninja108
+ * @copyright libnx Authors
+ */
+
+#include "services/nifm.h"
+#include "arm/atomics.h"
+
+typedef struct {
```
Comment (yellows8) [inline nx/source/services/nifm.c]: As said before, remove this and use Service directly.

**Comment by shadowninja108** [comment]
I thought you meant define IGeneralService in the source file since it wasn't needed in the header. Fixed.

File: nx/source/services/nifm.c [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+Result nifmInitialize(void) {
+    atomicIncrement64(&g_refCnt);
+
+    if (serviceIsActive(&g_nifmSrv))
+        return 0;
+		
+    Result rc;
+
+    rc = smGetService(&g_nifmSrv, "nifm:u");
+
```
Comment (yellows8) [inline nx/source/services/nifm.c]: Remove newline.

**Pull request merged**